sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,707 @@
+		if err != nil {
+			retErr = err
+			return
+		}
+		retErr = s.preloadBucketsFromDisk(ctx, diskBuckets)
+	})
+	if retErr != nil {
+		return nil, retErr
+	}
+
+	ret := make([]string, 0, 256)
+	s.bucketsCacheLock.RLock()
+	s.bucketsCache.Walk(func(s string, _ interface{}) bool {
+		ret = append(ret, s)
+		return false
+	})
+	s.bucketsCacheLock.RUnlock()
+
+	return ret, nil
+}
+
+// Parse a bucket from storage, but don't add it to the cache here.
+// (We may want to read a v1 bucket and upgrade it?)
+func (s *StoragePackerV2) DecodeBucket(storageEntry *logical.StorageEntry) (*LockedBucket, error) {
+	uncompressedData, notCompressed, err := compressutil.Decompress(storageEntry.Value)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to decompress packed storage entry: {{err}}", err)
+	}
+	if notCompressed {
+		uncompressedData = storageEntry.Value
+	}
+
+	var bucket Bucket
+	err = proto.Unmarshal(uncompressedData, &bucket)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to decode packed storage entry: {{err}}", err)
+	}
+
+	cacheKey := s.GetCacheKey(storageEntry.Key)
+	lock := locksutil.LockForKey(s.storageLocks, cacheKey)
+	lb := &LockedBucket{
+		LockEntry: lock,
+		Bucket:    &bucket,
+	}
+	if lb.Key == "" {
+		lb.Key = storageEntry.Key
+	}
+
+	return lb, nil
+}
+
+// Raw access to one of the storage buckets.
+// The caller is responsible for acquiring the lock and verifying that the bucket
+// has not been sharded in the meantime.
+func (s *StoragePackerV2) GetBucket(ctx context.Context, bucketKey string) (*LockedBucket, error) {
+	cacheKey := s.GetCacheKey(bucketKey)
+
+	s.bucketsCacheLock.RLock()
+	_, bucketRaw, found := s.bucketsCache.LongestPrefix(cacheKey)
+	s.bucketsCacheLock.RUnlock()
+
+	if found {
+		return bucketRaw.(*LockedBucket), nil
+	} else {
+		return nil, fmt.Errorf("key %q not found in cache", cacheKey)
+	}
+
+}
+
+func (s *StoragePackerV2) shardBucket(ctx context.Context, bucket *LockedBucket, cacheKey string) error {
+	numShards := int(math.Pow(2.0, float64(s.BucketShardBits)))
+
+	// Create the shards
+	s.Logger.Info("sharding bucket", "bucket_key", bucket.Key, "num_shards", numShards)
+	defer s.Logger.Info("sharding bucket process exited", "bucket_key", bucket.Key)
+
+	shards := make(map[string]*LockedBucket, numShards)
+	for i := 0; i < numShards; i++ {
+		shardKey := fmt.Sprintf("%x", i)
+		lock := locksutil.LockForKey(s.storageLocks, cacheKey+shardKey)
+		shardedBucket := &LockedBucket{
+			LockEntry: lock,
+			Bucket: &Bucket{
+				Key:       fmt.Sprintf("%s/%s", bucket.Key, shardKey),
+				ItemMap:   make(map[string][]byte),
+				HasShards: false,
+			},
+		}
+		shards[shardKey] = shardedBucket
+		s.Logger.Trace("created shard", "shard_key", shardKey)
+	}
+
+	s.Logger.Debug("resilvering items")
+
+	parentPrefix := s.GetCacheKey(bucket.Key)
+	// Resilver the items
+	for id, v := range bucket.ItemMap {
+		key := GetItemIDHash(id)
+		shardKey := strings.TrimPrefix(key, parentPrefix)[0 : s.BucketShardBits/4]
```

Comment:
Do we need to worry about panics here? I was [playing with this a little](https://play.golang.org/p/Qgv8lYQlyhm) and I can get it to panic in some cases, but I don't know if they represent real life or if we have guard rails around this elsewhere in the code.

Reply:
GetItemIDHash returns a 64-byte string, so it's theoretically but not practically possible to run off the end of the radix tree.

Let me think about what a check would look like; it probably wouldn't be here but it would be a limit on the maximum depth of the radix tree.

Reply:
I put in a check; if somebody has an odd sharding configuration this would be somewhat more feasible to create. (Perhaps we should have a check for too large of a BucketShardBits, setting it to a value like 32 or 64 could be very unpleasant.)